pkg/sql/backfill.go

```diff
@@ -1524,71 +1540,167 @@ func (e InvalidIndexesError) Error() string {
 	codec keys.SQLCodec,
+	db *kv.DB,
 	tableDesc catalog.TableDescriptor,
 	indexes []catalog.Index,
 	runHistoricalTxn sqlutil.HistoricalInternalExecTxnRunner,
 	withFirstMutationPublic bool,
 	gatherAllInvalid bool,
 	execOverride sessiondata.InternalExecutorOverride,
-) error {
+	protectedTSProvider protectedts.Provider,
+) (err error) {
+	protectedTSAdded := sync.Once{}
 	grp := ctxgroup.WithContext(ctx)
 	invalid := make(chan descpb.IndexID, len(indexes))
 
 	expectedCount := make([]int64, len(indexes))
 	countReady := make([]chan struct{}, len(indexes))
 
+	// Removes the protected timestamp, if one was added when this
+	// function returns.
+	var removeProtectedTS func(ctx context.Context) error
+	defer func() {
+		if removeProtectedTS != nil {
+			removeErr := removeProtectedTS(ctx)
+			if removeErr != nil {
+				if err != nil {
+					err = errors.WithSecondaryError(err, removeErr)
+				} else {
+					err = removeErr
+				}
+			}
+		}
+	}()
+
+	// Adds a protected timestamp only for historical queries.
+	var addProtectedTS = func() error {
+		var err error
+		protectedTSAdded.Do(func() {
+			// If we are not running a historical query, nothing to do here.
+			if runHistoricalTxn.ReadAsOf().IsEmpty() {
+				return
+			}
+			protectedtsID := uuid.MakeV4()
+			target := ptpb.MakeSchemaObjectsTarget(descpb.IDs{tableDesc.GetID()})
+			if err != nil {
+				return
+			}
+			rec := jobsprotectedts.MakeRecord(protectedtsID,
+				int64(jobID), runHistoricalTxn.ReadAsOf(), nil, jobsprotectedts.Jobs, target)
+			err = db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+				return protectedTSProvider.Protect(ctx, txn, rec)
+			})
+			if err != nil {
+				return
+			}
+			removeProtectedTS = func(ctx context.Context) error {
+				return db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+					return protectedTSProvider.Release(ctx, txn, protectedtsID)
+				})
+			}
+		})
+		return err
+	}
+
 	for i, idx := range indexes {
 		// Shadow i and idx to prevent the values from changing within each
 		// gorountine.
 		i, idx := i, idx
 		countReady[i] = make(chan struct{})
 
 		grp.GoCtx(func(ctx context.Context) error {
-			// KV scan can be used to get the index length.
-			// TODO (lucy): Switch to using DistSQL to get the count, so that we get
-			// distributed execution and avoid bypassing the SQL decoding
+
+			// Determine what the GC interval is on the table, which will help us
+			// figure out when to apply a protected timestamp, as a percentage of this
+			// time.
+			var waitBeforeProtectedTS time.Duration
 			start := timeutil.Now()
+			if err := runHistoricalTxn.Exec(ctx, func(ctx context.Context, txn *kv.Txn, _ sqlutil.InternalExecutor) error {
+				_, zoneCfg, _, err := GetZoneConfigInTxn(ctx, txn, codec, tableDesc.GetID(), nil, "", true)
+				if err != nil {
+					return err
+				}
+				waitBeforeProtectedTS = ((time.Duration(zoneCfg.GC.TTLSeconds) * time.Second) *
+					indexValidationProtectTimeStampGCPct) / 100
```

Comment:
For my own edification:

- If the zone configuration is changed after this lookup, will this wait time still be correct?

- Similarly, does running this in the historical transaction make sense?  Wouldn't we care about the most recent zone configuration?
-
- Is the cost of the protected timestamps high enough that we can't unconditionally install a PTS when starting the verification?  Is the idea that we don't want to pay that cost for small indexes that won't end up needing the PTS?